### PR TITLE
Preliminary zero-IQ port of VQF to Rust 

### PR DESCRIPTION
**Pull request by Noxime**
Its python. Like, it really is just python. This probably could have even been done with an AI. It's that dumb. Haven't tested it, might not work. Doesn't have the full API either. Straight up sucks

### CODE REVIEW

File: firmware/crates/vqf/.vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,3 @@
+{
+    "rust-analyzer.check.allTargets": false
+}
```
Comment (TheButlah) [inline firmware/crates/vqf/.vscode/settings.json]: this file shouldn't need to exist generally speaking

File: firmware/Cargo.toml [hunk 2]
```diff
@@ -261,7 +261,7 @@ opt-level = 'z'
 
 [workspace]
 resolver = "2"
-members = ["crates/panic_defmt", "crates/defmt_esp_println"]
+members = ["crates/panic_defmt", "crates/defmt_esp_println", "crates/vqf"]
```
Comment (TheButlah) [inline firmware/Cargo.toml]: vqf is not platform dependent, so it should not be in the firmware workspace. I would suggest putting it in the main workspace.

The basic strategy here is that as much code as possible is in the toplevel workspace, and only stuff that needs `.cargo/config.toml`, linker flags, and the like, should be in the firmware workspace.

**Comment by kitlith** [comment]
It would be nice to copy over the docs at some point, but that shouldn't be a blocker for merging and using it in the firmware, only for publishing for other people to use.

File: firmware/crates/vqf/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,11 @@
+[package]
+name = "vqf"
+version = "0.1.0"
+edition = "2021"
+license = "MIT"
```
Comment (TheButlah) [inline firmware/crates/vqf/Cargo.toml]: be  sure to copy the author's original MIT license file and paste it in the root of your crate.

File: firmware/crates/vqf/Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,11 @@
+[package]
+name = "vqf"
+version = "0.1.0"
```
Comment (TheButlah) [inline firmware/crates/vqf/Cargo.toml]: set this version to 0.0.0 for now, until we black box test against vqf-cxx. *then* we can publish to crates.io
```suggestion
version = "0.0.0"
```

File: firmware/crates/vqf/Cargo.toml (hunk 3 above)
Comment (TheButlah) [inline firmware/crates/vqf/Cargo.toml]:
```suggestion
license = "MIT"
publish = false
```

**Review by TheButlah — APPROVED** [review]
Awesome work! I would not have had the stomach for it myself. Glad someone did it.

We can use @kitlith's vqf-cxx crate to create some black box tests to compare this to the original implementation. Only once we do that should we think about renaming any of these functions. We can do that in a later PR though.

**Review by TheButlah — CHANGES_REQUESTED** [review]
.

**Comment by TheButlah** [comment]
build still fails

File: vqf/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,668 @@
...
+		// initialization phase
+		if state[(0, 1)].is_nan() {
+			// first sample
+			state[(0, 1)] = 0.0; // state[0, 1] is used to store the sample count
+					 // ඞ
+			state.get_mut((1, ..)).unwrap().fill(0.0); // state[1, :] is used to store the sum
+		}
+
+		state[(0, 1)] += 1.0;
+		let mut out = nalgebra::Matrix::zeros();
```
Comment (kitlith) [inline vqf/src/lib.rs]:
```suggestion
		let mut out = nalgebra::SMatrix::zeros();
```

File: vqf/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,668 @@
...
+	>,
+) -> nalgebra::Matrix<
+	f32,
+	nalgebra::Const<N>,
+	nalgebra::Const<1>,
+	ArrayStorage<f32, N, 1>,
+> {
+	// difference equations based on scipy.signal.lfilter documentation
+	// assumes that a0 == 1.0
+	let y = b[0] * x + nalgebra::Matrix::repeat(state[0]);
```
Comment (kitlith) [inline vqf/src/lib.rs]:
```suggestion
	let y = b[0] * x + nalgebra::SMatrix::repeat(state[0]);
```

**Review by kitlith — COMMENTED** [review]
These two suggested changes should fix CI. This changes from requesting a matrix of any kind to requesting specifically a statically sized matrix, explicitly excluding the possibility of a dynamically sized matrix.

(without alloc, there are no dynamically sized matrices, which is why the complaint only occurs when the std feature is enabled)

File: vqf/README.md [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+This crate implements the Versatile Quaternion-based Filter (VQF) for IMU Orientation Estimation from the publication
+
+> D. Laidig and T. Seel. "VQF: Highly Accurate IMU Orientation Estimation with Bias Estimation and Magnetic Disturbance Rejection." Information Fusion 2023, 91, 187--204. doi:10.1016/j.inffus.2022.10.014. [Accepted manuscript available at arXiv:2203.17024.]
+
+This code in its current form is a direct port of the Python code originally provided [here](https://github.com/dlaidig/vqf/)
+
+## License
+Copyright 2021 Daniel Laidig <laidig@control.tu-berlin.de>
+
+Licensed under the [MIT License](https://spdx.org/licenses/MIT.html)
```
Comment (TheButlah) [inline vqf/README.md:12]:
```suggestion
Licensed under the [MIT License](https://spdx.org/licenses/MIT.html)

All contributors to this crate additionally agree that should the original work be relicensed under a dual MIT/[Apache 2.0](https://spdx.org/licenses/Apache-2.0.html) license, their contributions will also be licensed this way.
```

**Review by TheButlah — CHANGES_REQUESTED** [review]
thoughts?

**Pull request merged**